src/config-array.js

```diff
@@ -443,6 +544,15 @@ export class ConfigArray extends Array {
 			this.push(configs);
 		}
 
+		/**
+		 * The number of base configs in the array.
+		 * @property baseConfigLength
+		 * @type {number}
+		 * @private
+		 * @readonly
+		 */
+		this[ConfigArraySymbol.baseConfigLength] = this.length;
```

Comment:
I could be missing something but it seems that `this.length` is always 0 at this point in code when ESLint is launched from the CLI. For the test case indicated in [this comment](https://github.com/humanwhocodes/config-array/pull/128#discussion_r1540851303) I'm still getting the error message `Error: Config "4": Unexpected key "foo" found.`.

Reply:
Yeah, in ESLint we're actually passing `configs` to the `ConfigArray` constructor and then inserting `baseConfig` (`defaultConfig`) afterwards.

https://github.com/eslint/eslint/blob/9b7bd3be066ac1f72fa35c4d31a1b178c7e2b683/lib/config/flat-config-array.js#L94-L108

I'm not sure why it was done this way. Perhaps we could refactor this to pass `baseConfig` to the `ConfigArray` constructor and then push provided `configs`.

Reply:
One thing that can be confusing is that by "base config" we mean the [default config](https://github.com/eslint/eslint/blob/main/lib/config/default-config.js). But there's also ESLint constructor option [`baseConfig`](https://github.com/eslint/eslint/blob/9b7bd3be066ac1f72fa35c4d31a1b178c7e2b683/lib/eslint/eslint.js#L71), which isn't used as a replacement for the default config but added between it and user-provided configs.

Reply:
I think we may be getting a bit too wrapped up in the specifics of how this works right now. The goal of this PR was really to expose `name` in an error message, and the index is just a fallback in case no `name` is specified. We can probably argue for a while about what is the correct fallback and how we may want to refactor `FlatConfigArray` to make things make more sense.

For now, I think the best choice is just to do what I had originally: use the actual index in the array and not try to distinguish between base and other configs. Realistically, the contents of a `ConfigArray` could change between instantiation and normalization, so any index is just a best guess.

How does that sound?

Reply:
Would it make sense to just rethrow the unwrapped `TypeError` when the config has no `name` property? We could follow up with a PR in ESLint to inform users that they are encouraged to add names to their configs in order to make debugging easier (with an info message on the terminal when a `TypeError` is detected).

Reply:
I agree with @fasttime. As it turns out that including the index in the error message wouldn't be very useful, and in fact rather confusing, it might be better not to include it and just use `name` if available.

Reply:
Okay, if the name isn't present, I'll use "(unknown)" instead to keep the error message format consistent. I'll also attach the index to the error object so we can maybe do something useful with that in ESLint.

Reply:
We could subtract 4 from the index in ESLint where the number of predefined config elements is known. But I understand from [this previous comment](https://github.com/humanwhocodes/config-array/pull/128#discussion_r1543431494) that the index is not guaranteed to be preserved during normalization?

Reply:
That's correct. Let's table any discussions around index for later. I don't think it's important right now, and the data will be available in ESLint if we want to use it for something.